Wikipedia/Code/TalkPageFetcher.swift

```diff
@@ -1,228 +1,55 @@
+import Foundation
+import WMF
 
-class NetworkTalkPage {
-    let url: URL
-    let topics: [NetworkTopic]
-    var revisionId: Int?
-    let displayTitle: String
-    
-    init(url: URL, topics: [NetworkTopic], revisionId: Int?, displayTitle: String) {
-        self.url = url
-        self.topics = topics
-        self.revisionId = revisionId
-        self.displayTitle = displayTitle
-    }
-}
-
-class NetworkBase: Codable {
-    let topics: [NetworkTopic]
-}
-
-class NetworkTopic:  NSObject, Codable {
-    let html: String
-    let replies: [NetworkReply]
-    let sectionID: Int
-    let shas: NetworkTopicShas
-    var sort: Int?
+struct TalkPageAPIResponse: Codable {
+    let threads: TalkPageThreadItems?
     
     enum CodingKeys: String, CodingKey {
-        case html
-        case shas
-        case replies
-        case sectionID = "id"
+        case threads = "discussiontoolspageinfo"
     }
 }
 
-class NetworkTopicShas: Codable {
-    let html: String
-    let indicator: String
-}
-
-class NetworkReply: NSObject, Codable {
-    let html: String
-    let depth: Int16
-    let sha: String
-    var sort: Int!
+struct TalkPageThreadItems: Codable {
+    let threadItems: [TalkPageItem]?
     
     enum CodingKeys: String, CodingKey {
-        case html
-        case depth
-        case sha
-    }
-}
-
-import Foundation
-import WMF
-
-enum TalkPageType: Int {
-    case user
-    case article
-    
-    func canonicalNamespacePrefix(for siteURL: URL) -> String? {
-        let namespace: PageNamespace
-        switch self {
-        case .article:
-            namespace = PageNamespace.talk
-        case .user:
-            namespace = PageNamespace.userTalk
-        }
-        return namespace.canonicalName + ":"
-    }
-    
-    func titleWithCanonicalNamespacePrefix(title: String, siteURL: URL) -> String {
-        return (canonicalNamespacePrefix(for: siteURL) ?? "") + title
-    }
-    
-    func titleWithoutNamespacePrefix(title: String) -> String {
-        if let firstColon = title.range(of: ":") {
-            var returnTitle = title
-            returnTitle.removeSubrange(title.startIndex..<firstColon.upperBound)
-            return returnTitle
-        } else {
-            return title
-        }
-    }
-    
-    func urlTitle(for title: String) -> String? {
-        assert(title.contains(":"), "Title must already be prefixed with namespace.")
-        return title.denormalizedPageTitle
+        case threadItems = "threaditemshtml"
     }
 }
 
-enum TalkPageFetcherError: Error {
-    case talkPageDoesNotExist
+struct TalkPageItem: Codable {
+    let type: String?
```

Comment:
Let's try removing some optionals in this struct.

1. Turn type into a non-optional String enum, with the raw values "comment" or "heading".
2. Let's remove `level` property entirely for now until we're sure we need it. As of now it just corresponds to the nesting level, so I'm not sure that we need it yet. It's also easy to confuse it with headingLevel (which I think of as more of a heading font size) so it would help to get rid of this one if we can.
3. Make `replies` non-optional.
4. It looks like the API sometimes spits out an `othercontent` String property here, so let's add an optional property for that. [Here's](https://test.wikipedia.org/w/api.php?action=discussiontoolspageinfo&format=json&page=Talk:Bird2&prop=threaditemshtml&formatversion=2) an example of one that does that. It's a more recent addition.

Unfortunately it seems like the other stuff could be optional (even `id`) from some of my API testing :(